src/App.tsx

```diff
@@ -1,24 +1,295 @@
+    if (!todosIdToDelete.length && !isLoadingAddTodo && formInput.current) {
+      formInput.current.focus();
+    }
+  }, [isLoadingAddTodo, todosIdToDelete]);
+
+  const handleChangeFilter = (newFilter: StatusEnum) => {
+    setFilter(newFilter);
+  };
+
+  const handleChangeNewTodoTitle
+    = (event: React.ChangeEvent<HTMLInputElement>) => {
+      setNewTodoTitle(event.target.value);
+    };
+
+  const hasCompletedTodo = useMemo(() => {
+    return todos.some(todo => todo.completed);
+  }, [todos]);
+
+  const notCompletedTodosLength = useMemo(() => {
+    return todos.filter(todo => !todo.completed).length;
+  }, [todos]);
+
+  const isCompletedAll = useMemo(() => {
+    return todos.every(todo => todo.completed);
+  }, [todos]);
+
+  const handelSubmit = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    if (!newTodoTitle.trim()) {
+      setErrorMessage('Title should not be empty');
+
+      return;
+    }
+
+    setErrorMessage('');
+    setIsLoadingAddTodo(true);
+
+    setTempTodo({
+      completed: false,
+      id: 0,
+      title: newTodoTitle.trim(),
+      userId: USER_ID,
+    });
+
+    setNewTodo({
+      userId: USER_ID,
+      title: newTodoTitle.trim(),
+      completed: false,
+    })
+      .then(response => {
+        setNewTodoTitle('');
+
+        return response;
+      })
+      .then((newTodo) => {
+        setTodos((prevState) => [...prevState, newTodo]);
+      })
+      .catch(() => {
+        setErrorMessage('Unable to add a todo');
+      })
+      .finally(() => {
+        setIsLoadingAddTodo(false);
+        setTempTodo(null);
+      });
+  };
+
+  useEffect(() => {
+    getTodos()
+      .then(setTodos)
+      .catch(() => {
+        setErrorMessage('Unable to load todos');
+      });
+  }, []);
+
+  const handelDeleteTodoById = (todoId: number) => {
+    setTodosIdToDelete(prevState => [...prevState, todoId]);
+
+    deleteTodoById(todoId)
+      .then(() => {
+        setTodos(prevState => prevState.filter(todo => todo.id !== todoId));
+      })
+      .catch(() => {
+        setErrorMessage('Unable to delete a todo');
+      })
+      .finally(() => {
+        setTodosIdToDelete(prevState => prevState.filter(id => id !== todoId));
+      });
+  };
+
+  const handleChangeCompletedStatus
+    = (todoId: number, isCompleted: boolean) => {
+      setTodosIdToUpdate(prevState => [...prevState, todoId]);
+
+      updateTodoById(todoId, { completed: !isCompleted })
+        .then(() => {
+          setTodos(prevState => {
+            const currentTodo = prevState
+              .find(todo => todo.id === todoId);
+
+            if (!currentTodo) {
+              return prevState;
+            }
+
+            const newTodos = [...prevState];
+            const index = prevState.indexOf(currentTodo);
+
+            newTodos.splice(index, 1, {
+              ...currentTodo,
+              completed: !currentTodo.completed,
+            });
+
+            return newTodos;
```

Comment:
Just use `map` instead of `indexOf` + `splice` and `find` we don't need in this case too